src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -1584,38 +1585,20 @@ protected RubyNode coerceToBoolean(RubyNode includeProtectedAndPrivate) {
         }
 
         @Specialization
-        protected boolean doesRespondToString(
+        protected boolean doesRespondTo(
                 VirtualFrame frame,
                 Object object,
-                RubyString name,
+                Object name,
                 boolean includeProtectedAndPrivate,
-                @Cached ToJavaStringNode toJavaString) {
-            final boolean ret;
-            useCallerRefinements(frame);
-
-            if (ignoreVisibilityProfile.profile(includeProtectedAndPrivate)) {
-                ret = dispatchIgnoreVisibility.execute(frame, object, toJavaString.executeToJavaString(name));
-            } else {
-                ret = dispatch.execute(frame, object, toJavaString.executeToJavaString(name));
-            }
-
-            if (isTrueProfile.profile(ret)) {
-                return true;
-            } else if (respondToMissingProfile
-                    .profile(dispatchRespondToMissing.execute(frame, object, "respond_to_missing?"))) {
-                return respondToMissing(object, getSymbol(name.rope), includeProtectedAndPrivate);
-            } else {
-                return false;
+                @Cached ConditionProfile notSymbolOrStringProfile,
+                @Cached ToJavaStringNode toJavaString,
+                @Cached ToSymbolNode toSymbolNode) {
+            if (notSymbolOrStringProfile.profile(!RubyGuards.isRubySymbolOrString(name))) {
+                throw new RaiseException(
+                        getContext(),
+                        coreExceptions().typeErrorIsNotAOrB(object.toString(), "symbol", "string", this));
```

Comment:
`toString()` should not be called in PE code, because the receiver class is not known so it would be a virtual call, which is a performance warning.
So I'd suggest passing `object` to `typeErrorIsNotAOrB()` (then toString() is called behind a TruffleBoundary and that's OK).

We should probably use the new `CoreException#receiverString()` helper there instead of toString(), but the PR adding it is not merged yet (should be soon).

Reply:
Hm, it seems we have a few places with similar`.toString()`. 
https://github.com/oracle/truffleruby/blob/314f8ae9671d7b54aedbeb43194a64cf92497cff/src/main/java/org/truffleruby/core/cast/LongCastNode.java#L47-L51

and

https://github.com/oracle/truffleruby/blob/4acd2a83701ab13d68c2b39531052d7409968bd7/src/main/java/org/truffleruby/core/cast/BigIntegerCastNode.java#L68-L70

here we use `Utils.toString()`:

https://github.com/oracle/truffleruby/blob/314f8ae9671d7b54aedbeb43194a64cf92497cff/src/main/java/org/truffleruby/core/cast/IntegerCastNode.java#L54-L58

Is this will be a part of 
>We should probably use the new CoreException#inspectReceiver() helper there instead of toString(), but the PR adding it is not merged yet (should be soon).

or we need to refactor these places too (probably not in this PR)?

Reply:
Those cases are behind a `@TruffleBoundary`, so compilation-wise they are fine. The case here isn't.
I think `-H:+TruffleCheckBlackListedMethods` which is used in `jt build --env native` would actually detect as an error at image build time.

In terms of error message, inspectReceiver() is better. I used that to replace `toString()` in CoreExceptions, but not in other places (yet).

Reply:
The PR I was talking about has been merged now, so if you rebase you can use `inspectReceiver()`:
https://github.com/oracle/truffleruby/commit/25eacca413e7acb3816be73199cf58f865510279
As usual, new code should use the correct approach, but fixing old code is out of scope.

Reply:
> Those cases are behind a @TruffleBoundary, so compilation-wise they are fine

Ah, that was obvious, sorry

> The PR I was talking about has been merged now, so if you rebase you can use inspectReceiver():

No problem, but I should still pass `Object` over `String`, right?:

```java
public RubyException typeErrorIsNotAOrB(Object value, String expectedTypeA, String expectedTypeB, Node currentNode) {
   # use inspectReceiver(value) instead of value.toString()
```

Reply:
Yes, I think in the future we should remove those cases were the receiver is typed as `String`, conversation should be done in CoreExeptions or later.